test_rmw_implementation/test/test_service.cpp

```diff
@@ -0,0 +1,186 @@
+  rmw_qos_profile_t native_qos_profile = rmw_qos_profile_default;
+  native_qos_profile.avoid_ros_namespace_conventions = true;
+  rmw_service_t * srv =
+    rmw_create_service(node, ts, service_name, &native_qos_profile);
+  ASSERT_NE(nullptr, srv) << rmw_get_error_string().str;
+  rmw_ret_t ret = rmw_destroy_service(node, srv);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+}
+
+TEST_F(CLASSNAME(TestService, RMW_IMPLEMENTATION), create_with_bad_arguments) {
+  constexpr char service_name[] = "/test";
+  const rosidl_service_type_support_t * ts =
+    ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, srv, BasicTypes);
+
+  rmw_service_t * srv =
+    rmw_create_service(nullptr, ts, service_name, &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, srv);
+  rmw_reset_error();
+
+  srv = rmw_create_service(node, nullptr, service_name, &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, srv);
+  rmw_reset_error();
+
+  const char * implementation_identifier = node->implementation_identifier;
+  node->implementation_identifier = "not-an-rmw-implementation-identifier";
+  srv = rmw_create_service(node, ts, service_name, &rmw_qos_profile_default);
+  node->implementation_identifier = implementation_identifier;
+  EXPECT_EQ(nullptr, srv);
+  rmw_reset_error();
+
+  srv = rmw_create_service(node, ts, nullptr, &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, srv);
+  rmw_reset_error();
+
+  srv = rmw_create_service(node, ts, "", &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, srv);
+  rmw_reset_error();
+
+  constexpr char service_name_with_spaces[] = "/foo bar";
+  srv = rmw_create_service(node, ts, service_name_with_spaces, &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, srv);
+  rmw_reset_error();
+
+  constexpr char relative_service_name[] = "foo";
+  srv = rmw_create_service(node, ts, relative_service_name, &rmw_qos_profile_default);
+  EXPECT_EQ(nullptr, srv);
+  rmw_reset_error();
+
+  srv = rmw_create_service(node, ts, service_name, nullptr);
+  EXPECT_EQ(nullptr, srv);
+  rmw_reset_error();
+
+  srv = rmw_create_service(node, ts, service_name, &rmw_qos_profile_unknown);
+  EXPECT_EQ(nullptr, srv);
+  rmw_reset_error();
+
+  // Creating and destroying a service still succeeds.
+  srv = rmw_create_service(node, ts, service_name, &rmw_qos_profile_default);
+  ASSERT_NE(nullptr, srv) << rmw_get_error_string().str;
+  rmw_ret_t ret = rmw_destroy_service(node, srv);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+}
+
+TEST_F(CLASSNAME(TestService, RMW_IMPLEMENTATION), destroy_with_bad_arguments) {
+  constexpr char service_name[] = "/test";
+  const rosidl_service_type_support_t * ts =
+    ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, srv, BasicTypes);
+  rmw_service_t * srv =
+    rmw_create_service(node, ts, service_name, &rmw_qos_profile_default);
+  ASSERT_NE(nullptr, srv) << rmw_get_error_string().str;
+
+  // Destroying service with invalid arguments fails.
+  rmw_ret_t ret = rmw_destroy_service(nullptr, srv);
```

Comment:
If this succeeds, couldn't the failure bleed into the other checks below? Should you be creating a valid service every time?

Reply:
Failures would bleed out, yes, and likely cause a crash, as the service has been freed. 

We _could_ split each bad destruction attempt into its own test, and assert that the destruction call did not succeed before proceeding to destruct it (which is part of the test). But either way we have to trust `rmw_destroy_service()` outcome. There's no other way to test for validity.

Would you have rather have N tests? I don't feel strongly about it.

Reply:
I do like N `TEST_F`s a bit better, especially if you created a small test fixture. I'm realizing I probably did something like what you've done here plenty of times, but more recently, I've been leaning towards more `TEST_F` rather than fewer so issues in failing tests can be isolated quickly.

Reply:
Fair enough. See https://github.com/ros2/rmw_implementation/pull/138/commits/8680b3ea9182f19115ce613d7022dd4d4a398a2a.